### PR TITLE
Implement redeem voucher service

### PR DESCRIPTION
**Pull request by mojganii**
Implement redeeming the voucher code

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747)

### CODE REVIEW

File: ios/MullvadREST/RESTAPIProxy.swift [hunk 1]
```diff
@@ -208,6 +208,50 @@ extension REST {
...
+                },
+                authorizationProvider: createAuthorizationProvider(accountNumber: accountNumber)
+            )
+
+            let responseHandler = AnyResponseHandler { response, data -> ResponseHandlerResult<SubmitVoucherResponse> in
+                if HTTPStatus.isSuccess(response.statusCode) {
+                    return .decoding {
+                        try self.responseDecoder.decode(SubmitVoucherResponse.self, from: data)
+                    }
+                } else {
```
Comment (MrChocolatine) [inline ios/MullvadREST/RESTAPIProxy.swift:240]: This `else` branch seems extraneous due to the `if return`.

**Review by mojganii — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @MrChocolatine)

___
*[`ios/MullvadREST/RESTAPIProxy.swift` line 240 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#gh-1210251869:-NWh9p8x9wlO1DFOMSfM:b-ffreo5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/c9fea68361b586118901090bc8148df0f4e2da5e/ios/MullvadREST/RESTAPIProxy.swift#L240)):*
<details><summary><i>Previously, MrChocolatine wrote…</i></summary><blockquote>

This `else` branch seems extraneous due to the `if return`.
</blockquote></details>

what do you mean?

**Comment by MrChocolatine** [comment]
> > _[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747)_ status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @MrChocolatine)
> 
> _[`ios/MullvadREST/RESTAPIProxy.swift` line 240 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#gh-1210251869:-NWh9p8x9wlO1DFOMSfM:b-ffreo5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/c9fea68361b586118901090bc8148df0f4e2da5e/ios/MullvadREST/RESTAPIProxy.swift#L240)):_
> 
> Previously, MrChocolatine wrote…
> > This `else` branch seems extraneous due to the `if return`.
> 
> what do you mean?

I see my comment does not apply anymore as the code was cleaned :-) .

For the record, I simply meant, going from this:

```swift
if x {
  return y
} else {
  return z
}
```

to this:

```swift
if x {
  return y
}

return z
```

**Review by Tropical-Smootie — COMMENTED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747)* status: all files reviewed, 1 unresolved discussion (waiting on @MrChocolatine)

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> Dismissed @MrChocolatine from a discussion.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747)* status: all files reviewed, 3 unresolved discussions (waiting on @mojganii)

___
*[`ios/MullvadREST/RESTAPIProxy.swift` line 240 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#gh-1210251869:-NXLvVjB7N-R-wfAt1QY:b-xxbga) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/c9fea68361b586118901090bc8148df0f4e2da5e/ios/MullvadREST/RESTAPIProxy.swift#L240)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

what do you mean?
</blockquote></details>

Seems like this comment is out of context.
___
*[`ios/MullvadREST/RESTAPIProxy.swift` line 236 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#-NXLvam3BsXGadCAHubh:-NXLvam3BsXGadCAHubi:b6mg1la) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/56fb77a5cc081101b4307e6192d30fe63cc204ae/ios/MullvadREST/RESTAPIProxy.swift#L236)):*
> ```Swift
> 
>             let responseHandler = AnyResponseHandler { response, data -> ResponseHandlerResult<SubmitVoucherResponse> in
>                 guard HTTPStatus.isSuccess(response.statusCode) else {
> ```

nit: a bit more straight forward to use if-else in simple conditions than guards, i.e:

```
guard <condition> else {
  return <else_value>
}
return <then_value>
```

vs.

```
if <condition> {
  return <then_value>
} else {
  return <else_value>
}
```
___
*[`ios/MullvadREST/RESTAPIProxy.swift` line 247 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#-NXLtHtZ8pR4musUa2A1:-NXLtHtZ8pR4musUa2A2:bfen85j) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/56fb77a5cc081101b4307e6192d30fe63cc204ae/ios/MullvadREST/RESTAPIProxy.swift#L247)):*
> ```Swift
> 
>             return addOperation(
>                 name: "send-submit-voucher",
> ```

Could you please stick to the actual name of the REST call, i.e `submit-voucher` so it's easier to follow. Both send and submit are synonymous so when both are put in a row it looks a bit odd.

**Review by pronebird — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4747#-:-NXRJXOiBZVQuTD6hLDO:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**